src/tango_sdp_subarray/SDPSubarray/SDPSubarray.py

```diff
@@ -268,40 +346,400 @@ def ConfigureScan(self, scan_config, schema_path=None):
         pb_config = json.loads(scan_config)
         validate(pb_config, schema)
 
+        # Update receive addresses (if required) for the new scan configuration
+        # self._update_receive_addresses()
+
+        LOG.debug('Setting ObsState to READY.')
         self._obs_state = ObsState.READY
 
     @command
     @DebugIt()
     def StartScan(self):
         """Command issued when a scan is started."""
+        self._require_obs_state([ObsState.READY])
         self._obs_state = ObsState.SCANNING
 
     @command
     @DebugIt()
     def EndScan(self):
         """Command issued when the scan is ended."""
+        self._require_obs_state([ObsState.SCANNING])
         self._obs_state = ObsState.READY
 
     @command
     @DebugIt()
     def EndSB(self):
         """Command issued to end the scheduling block."""
+        self._require_obs_state([ObsState.READY])
         self._obs_state = ObsState.IDLE
 
     def _scan_complete(self):
         """Update the obsState to READY when a scan is complete.
 
         Internal state transition.
+
         """
+        self._require_obs_state([ObsState.SCANNING])
         self._obs_state = ObsState.READY
 
+    def _evaluate_receive_host_channel_map(self):
+        """Evaluate channel map for receive hosts.
+
+        NOTE(BMo) This is a placeholder function and will need completely
+                  reworking at some point!
+
+        This function should generate a map of receive hosts to
+        channels. This will eventually need to depend on the deployment
+        of the receive EE processes.
+
+        At the moment the PB config parameters for the vis_ingest
+        example are not sufficient to do this property.
+
+        """
+        if self._pb_config['workflow']['id'] != 'vis_ingest':
+            return
+
+        LOG.debug('Evaluating receive host channel map.')
+        self._recv_hosts = list()
+
+        # FIXME(BMo) Validate against expected PB parameters!
+
+        pb_params = self._pb_config['parameters']
+        num_channels = pb_params['numChannels']
+
+        # FIXME(BMo) parameter does not yet exist... default is always used!
+        max_channels_per_host = pb_params.get('maxChannelsPerHost', 400)
+        num_hosts = ceil(num_channels / max_channels_per_host)
+
+        LOG.debug('No. channels: %d', num_channels)
+        LOG.debug('Max channels per host: %d', max_channels_per_host)
+        LOG.debug('No. hosts: %d', num_hosts)
+
+        # FIXME(BMo) complete hack until we have a way of getting actual
+        #            receive hosts!
+        for host_index in range(num_hosts):
+            host_ip = '192.168.{}.{}'.format(
+                host_index//256,
+                host_index - ((host_index//256) * 256) + 1)
```

Comment:
Not that it means anything, but aren't we starting at 192.168.0.-1 here?

Reply:
nope starting at `192.168.0.1` for `host_index == 0`. I had to double check though as the brackets are a bit confusing.
There is another bug here though i've just fixed that i only just noticed when looking at this. For `host_index == 255` this code generates an IP of `192.168.0.256` which is invalid! 

I had added the `+1` to try to start at ``192.168.0.1`, clearly a bad idea!

Anyway this is fixed now in the version i'm about to push

Reply:
yes also it means nothing as this code should get replaced when we have time.